File: docdoc/qrbill.py
"""Swiss QR-bill: decode QR codes from page images (zbarimg) and parse the
Swiss Payments Code payload.

Implements SIX Implementation Guidelines v2.3 (accepting v2.2 'K' combined
addresses too -- pre-Nov-2025 paper is everywhere) and Swico S1 billing
information. Payload Version is '0200' for all of v2.0-2.4, so spec version
cannot be detected from the payload.
"""

import datetime
import re
import subprocess

QR_IID_MIN, QR_IID_MAX = 30000, 31999
MOD10_TABLE = [0, 9, 4, 6, 8, 2, 7, 1, 3, 5]

NOTIFICATION_TEXTS = (
    "NICHT ZUR ZAHLUNG VERWENDEN",
    "NE PAS UTILISER POUR LE PAIEMENT",
    "NON UTILIZZARE PER IL PAGAMENTO",
    "DO NOT USE FOR PAYMENT",
)


def iban_valid(iban):
    s = re.sub(r"\s", "", iban or "").upper()
    if not re.fullmatch(r"[A-Z]{2}[0-9]{2}[A-Z0-9]+", s):
        return False
    rearranged = s[4:] + s[:4]
    digits = "".join(str(int(c, 36)) for c in rearranged)
    return int(digits) % 97 == 1


def is_qr_iban(iban):
    s = re.sub(r"\s", "", iban or "")
    if len(s) < 9:
        return False
    try:
        iid = int(s[4:9])
    except ValueError:
        return False
    return QR_IID_MIN <= iid <= QR_IID_MAX


def qrr_valid(ref):
    """QR reference: 27 digits, mod-10 recursive check digit, not all zeros."""
    if not re.fullmatch(r"\d{27}", ref or "") or ref.strip("0") == "":
        return False
    carry = 0
    for d in ref[:26]:
        carry = MOD10_TABLE[(carry + int(d)) % 10]
    return (10 - carry) % 10 == int(ref[26])


def scor_valid(ref):
    """ISO 11649 creditor reference: RF + 2 check digits + <=21 alnum."""
    s = (ref or "").upper()
    if not re.fullmatch(r"RF\d{2}[A-Z0-9]{1,21}", s):
        return False
    rearranged = s[4:] + s[:4]
    digits = "".join(str(int(c, 36)) for c in rearranged)
    return int(digits) % 97 == 1


def _swico_split(s):
    """Split on '/' honoring \\/ and \\\\ escapes."""
    parts, cur, i = [], "", 0
    while i < len(s):
        c = s[i]
        if c == "\\" and i + 1 < len(s) and s[i + 1] in "/\\":
            cur += s[i + 1]
            i += 2
        elif c == "/":
            parts.append(cur)
            cur = ""
            i += 1
        else:
            cur += c
            i += 1
    parts.append(cur)
    return parts


def _yymmdd(s):
    try:
        d = datetime.datetime.strptime(s[:6], "%y%m%d").date()
        return d.isoformat()
    except ValueError:
        return None


def parse_swico(strd):
    """Parse Swico S1 billing information ('//S1/10/.../40/2:10;0:30').

    Returns dict with raw tags plus derived: invoice_no, invoice_date,
    customer_ref, uid, due_date (from /11/ + the 0%-entry of /40/),
    discounts [(pct, days), ...].
    """
    if not strd or not strd.startswith("//"):
        return None
    scheme = strd[2:4]
    if scheme != "S1":
        return {"scheme": scheme, "raw": strd}
    parts = _swico_split(strd[4:])
    # parts[0] is '' (leading /), then alternating tag, value
    tags = {}
    it = iter(parts[1:])
    for tag in it:
        val = next(it, "")
        if tag:
            tags[tag] = val
    out = {"scheme": "S1", "tags": tags, "raw": strd}
    out["invoice_no"] = tags.get("10")
    out["invoice_date"] = _yymmdd(tags["11"]) if "11" in tags else None
    out["customer_ref"] = tags.get("20")
    out["uid"] = tags.get("30")
    if "40" in tags:
        discounts = []
        for cond in tags["40"].split(";"):
            m = re.fullmatch(r"([\d.]+):(\d+)", cond.strip())
            if m:
                discounts.append((float(m.group(1)), int(m.group(2))))
        out["discounts"] = discounts
        net = next((days for pct, days in discounts if pct == 0), None)
        if net is not None and out["invoice_date"]:
            due = (datetime.date.fromisoformat(out["invoice_date"])
                   + datetime.timedelta(days=net))
            out["due_date"] = due.isoformat()
    return out


def _address(lines):
    """7 payload lines -> address dict (S structured or K combined)."""
    adrtp, name, l1, l2, pcode, town, country = (lines + [""] * 7)[:7]
    if not name:
        return None
    a = {"name": name, "country": country or None, "type": adrtp or None}
    if adrtp == "K":
        a["line1"], a["line2"] = l1 or None, l2 or None
    else:
        a["street"], a["building"] = l1 or None, l2 or None
        a["postal_code"], a["town"] = pcode or None, town or None
    return a


def parse_spc(payload):
    """Parse a Swiss Payments Code payload. Returns dict or None if not SPC.

    Tolerant reader: validation problems recorded in ['problems'] instead of
    rejecting (scanned archives contain old and slightly broken bills).
    """
    if not payload:
        return None
    lines = payload.replace("\r\n", "\n").split("\n")
    if len(lines) < 31 or lines[0].strip() != "SPC":
        return None
    L = [ln.strip() for ln in lines] + [""] * (34 - len(lines))
    problems = []
    if L[1] != "0200":
        problems.append(f"unexpected version {L[1]!r}")
    if L[2] != "1":
        problems.append(f"unexpected coding {L[2]!r}")

    iban = L[3].replace(" ", "").upper()
    if not iban_valid(iban):
        problems.append(f"IBAN check failed: {iban}")
    if iban[:2] not in ("CH", "LI"):
        problems.append(f"IBAN not CH/LI: {iban[:2]}")

    creditor = _address(L[4:11])
    debtor = _address(L[20:27])

    amount = None
    if L[18]:
        try:
            amount = float(L[18])
        except ValueError:
            problems.append(f"bad amount {L[18]!r}")
    currency = L[19]
    if currency not in ("CHF", "EUR"):
        problems.append(f"bad currency {currency!r}")

    ref_type, reference = L[27], L[28].replace(" ", "")
    if ref_type == "QRR":
        if not qrr_valid(reference):
            problems.append(f"QRR reference invalid: {reference}")
        if not is_qr_iban(iban):
            problems.append("QRR used with non-QR-IBAN")
    elif ref_type == "SCOR":
        if not scor_valid(reference):
            problems.append(f"SCOR reference invalid: {reference}")
        if is_qr_iban(iban):
            problems.append("SCOR used with QR-IBAN")
    elif ref_type == "NON":
        if reference:
            problems.append("reference present with type NON")
    else:
        problems.append(f"unknown reference type {ref_type!r}")

    message = L[29]
    if L[30] != "EPD":
        problems.append(f"missing EPD trailer (got {L[30]!r})")
    swico = parse_swico(L[31]) if L[31] else None
    alt = [x for x in L[32:34] if x]

    is_notification = (amount == 0.0 or
                       any(t in (message or "").upper() for t in NOTIFICATION_TEXTS))

    return {
        "iban": iban,
        "is_qr_iban": is_qr_iban(iban),
        "creditor": creditor,
        "debtor": debtor,
        "amount": amount,
        "currency": currency,
        "ref_type": ref_type,
        "reference": reference or None,
        "message": message or None,
        "swico": swico,
        "alt_procedures": alt,
        "is_notification": is_notification,
        "problems": problems,
        "payload": payload,
    }


def decode_qr_codes(image_path):
    """All QR payloads in an image via zbarimg. Returns list of strings."""
    proc = subprocess.run(
        ["zbarimg", "--raw", "-q", "-Sdisable", "-Sqrcode.enable", "-Sbinary",
         image_path],
        capture_output=True)
    if proc.returncode not in (0, 4):        # 4 = no symbols found
        return []
    # -Sbinary stops zbar from charset-guessing (it mangles UTF-8 umlauts
    # otherwise); we decode UTF-8 ourselves. SPC payloads embed newlines,
    # so symbols are split on the 'SPC' header anchor, not on lines.
    out = proc.stdout.decode("utf-8", errors="replace")
    if not out.strip():
        return []
    idxs = [m.start() for m in re.finditer(r"(?m)^SPC\r?$", out)]
    if idxs:
        idxs.append(len(out))
        return [out[idxs[i]:idxs[i + 1]].strip("\n") for i in range(len(idxs) - 1)]
    return [c for c in out.strip("\n").split("\n") if c]


def find_qrbill(image_paths):
    """Scan page images for a Swiss QR-bill; first valid SPC wins."""
    for path in image_paths:
        for payload in decode_qr_codes(path):
            parsed = parse_spc(payload)
            if parsed:
                return parsed
    return None


def build_spc(qr):
    """Rebuild an SPC payload from a parsed dict (for QR re-render).
    Uses the stored original payload when available."""
    return qr.get("payload")

File: app/lib/qrbill.js
// Swiss QR-bill: decode QR codes from page images (zbarimg) and parse the
// Swiss Payments Code payload.
//
// Implements SIX Implementation Guidelines v2.3 (accepting v2.2 'K' combined
// addresses too -- pre-Nov-2025 paper is everywhere) and Swico S1 billing
// information. Payload Version is '0200' for all of v2.0-2.4, so spec version
// cannot be detected from the payload.

const { execFile } = require("child_process");
const { promisify } = require("util");
const execFileP = promisify(execFile);

const QR_IID_MIN = 30000, QR_IID_MAX = 31999;
const MOD10_TABLE = [0, 9, 4, 6, 8, 2, 7, 1, 3, 5];

const NOTIFICATION_TEXTS = [
  "NICHT ZUR ZAHLUNG VERWENDEN",
  "NE PAS UTILISER POUR LE PAIEMENT",
  "NON UTILIZZARE PER IL PAGAMENTO",
  "DO NOT USE FOR PAYMENT",
];

// big-int mod 97 over a digit string (IBAN/SCOR check)
function mod97(digits) {
  let rem = 0;
  for (const ch of digits) rem = (rem * 10 + (ch.charCodeAt(0) - 48)) % 97;
  return rem;
}

const alnumDigits = (s) =>
  [...s].map((c) => parseInt(c, 36).toString()).join("");

function ibanValid(iban) {
  const s = String(iban || "").replace(/\s/g, "").toUpperCase();
  if (!/^[A-Z]{2}[0-9]{2}[A-Z0-9]+$/.test(s)) return false;
  return mod97(alnumDigits(s.slice(4) + s.slice(0, 4))) === 1;
}

function isQrIban(iban) {
  const s = String(iban || "").replace(/\s/g, "");
  if (s.length < 9) return false;
  const iid = parseInt(s.slice(4, 9), 10);
  return Number.isInteger(iid) && iid >= QR_IID_MIN && iid <= QR_IID_MAX;
}

function qrrValid(ref) {
  // QR reference: 27 digits, mod-10 recursive check digit, not all zeros.
  if (!/^\d{27}$/.test(ref || "") || /^0+$/.test(ref)) return false;
  let carry = 0;
  for (const d of ref.slice(0, 26))
    carry = MOD10_TABLE[(carry + (d.charCodeAt(0) - 48)) % 10];
  return (10 - carry) % 10 === parseInt(ref[26], 10);
}

function scorValid(ref) {
  // ISO 11649 creditor reference: RF + 2 check digits + <=21 alnum.
  const s = String(ref || "").toUpperCase();
  if (!/^RF\d{2}[A-Z0-9]{1,21}$/.test(s)) return false;
  return mod97(alnumDigits(s.slice(4) + s.slice(0, 4))) === 1;
}

function swicoSplit(s) {
  // Split on '/' honoring \/ and \\ escapes.
  const parts = [];
  let cur = "", i = 0;
  while (i < s.length) {
    const c = s[i];
    if (c === "\\" && i + 1 < s.length && "/\\".includes(s[i + 1])) {
      cur += s[i + 1]; i += 2;
    } else if (c === "/") {
      parts.push(cur); cur = ""; i += 1;
    } else {
      cur += c; i += 1;
    }
  }
  parts.push(cur);
  return parts;
}

function yymmdd(s) {
  const m = /^(\d{2})(\d{2})(\d{2})/.exec(s || "");
  if (!m) return null;
  const [, yy, mo, dd] = m;
  const year = 2000 + parseInt(yy, 10);
  const d = new Date(Date.UTC(year, parseInt(mo, 10) - 1, parseInt(dd, 10)));
  if (d.getUTCMonth() + 1 !== parseInt(mo, 10) || d.getUTCDate() !== parseInt(dd, 10))
    return null;
  return d.toISOString().slice(0, 10);
}

const addDays = (iso, days) => {
  const d = new Date(iso + "T00:00:00Z");
  d.setUTCDate(d.getUTCDate() + days);
  return d.toISOString().slice(0, 10);
};

function parseSwico(strd) {
  // Parse Swico S1 billing information ('//S1/10/.../40/2:10;0:30').
  // Returns dict with raw tags plus derived: invoice_no, invoice_date,
  // customer_ref, uid, due_date (from /11/ + the 0%-entry of /40/),
  // discounts [[pct, days], ...].
  if (!strd || !strd.startsWith("//")) return null;
  const scheme = strd.slice(2, 4);
  if (scheme !== "S1") return { scheme, raw: strd };
  const parts = swicoSplit(strd.slice(4));
  // parts[0] is '' (leading /), then alternating tag, value
  const tags = {};
  for (let i = 1; i < parts.length; i += 2)
    if (parts[i]) tags[parts[i]] = parts[i + 1] ?? "";
  const out = { scheme: "S1", tags, raw: strd };
  out.invoice_no = tags["10"] ?? null;
  out.invoice_date = "11" in tags ? yymmdd(tags["11"]) : null;
  out.customer_ref = tags["20"] ?? null;
  out.uid = tags["30"] ?? null;
  if ("40" in tags) {
    const discounts = [];
    for (const cond of tags["40"].split(";")) {
      const m = /^([\d.]+):(\d+)$/.exec(cond.trim());
      if (m) discounts.push([parseFloat(m[1]), parseInt(m[2], 10)]);
    }
    out.discounts = discounts;
    const net = discounts.find(([pct]) => pct === 0)?.[1];
    if (net !== undefined && out.invoice_date)
      out.due_date = addDays(out.invoice_date, net);
  }
  return out;
}

function address(lines) {
  // 7 payload lines -> address dict (S structured or K combined).
  const [adrtp, name, l1, l2, pcode, town, country] =
    [...lines, "", "", "", "", "", "", ""].slice(0, 7);
  if (!name) return null;
  const a = { name, country: country || null, type: adrtp || null };
  if (adrtp === "K") {
    a.line1 = l1 || null; a.line2 = l2 || null;
  } else {
    a.street = l1 || null; a.building = l2 || null;
    a.postal_code = pcode || null; a.town = town || null;
  }
  return a;
}

function parseSpc(payload) {
  // Parse a Swiss Payments Code payload. Returns dict or null if not SPC.
  // Tolerant reader: validation problems recorded in .problems instead of
  // rejecting (scanned archives contain old and slightly broken bills).
  if (!payload) return null;
  const lines = payload.replace(/\r\n/g, "\n").split("\n");
  if (lines.length < 31 || lines[0].trim() !== "SPC") return null;
  const L = lines.map((ln) => ln.trim());
  while (L.length < 34) L.push("");
  const problems = [];
  if (L[1] !== "0200") problems.push(`unexpected version '${L[1]}'`);
  if (L[2] !== "1") problems.push(`unexpected coding '${L[2]}'`);

  const iban = L[3].replace(/\s/g, "").toUpperCase();
  if (!ibanValid(iban)) problems.push(`IBAN check failed: ${iban}`);
  if (!["CH", "LI"].includes(iban.slice(0, 2)))
    problems.push(`IBAN not CH/LI: ${iban.slice(0, 2)}`);

  const creditor = address(L.slice(4, 11));
  const debtor = address(L.slice(20, 27));

  let amount = null;
  if (L[18]) {
    amount = parseFloat(L[18]);
    if (!Number.isFinite(amount)) {
      problems.push(`bad amount '${L[18]}'`);
      amount = null;
    }
  }
  const currency = L[19];
  if (!["CHF", "EUR"].includes(currency))
    problems.push(`bad currency '${currency}'`);

  const refType = L[27], reference = L[28].replace(/\s/g, "");
  if (refType === "QRR") {
    if (!qrrValid(reference)) problems.push(`QRR reference invalid: ${reference}`);
    if (!isQrIban(iban)) problems.push("QRR used with non-QR-IBAN");
  } else if (refType === "SCOR") {
    if (!scorValid(reference)) problems.push(`SCOR reference invalid: ${reference}`);
    if (isQrIban(iban)) problems.push("SCOR used with QR-IBAN");
  } else if (refType === "NON") {
    if (reference) problems.push("reference present with type NON");
  } else {
    problems.push(`unknown reference type '${refType}'`);
  }

  const message = L[29];
  if (L[30] !== "EPD") problems.push(`missing EPD trailer (got '${L[30]}')`);
  const swico = L[31] ? parseSwico(L[31]) : null;
  const alt = L.slice(32, 34).filter(Boolean);

  const isNotification = amount === 0.0 ||
    NOTIFICATION_TEXTS.some((t) => (message || "").toUpperCase().includes(t));

  return {
    iban,
    is_qr_iban: isQrIban(iban),
    creditor,
    debtor,
    amount,
    currency,
    ref_type: refType,
    reference: reference || null,
    message: message || null,
    swico,
    alt_procedures: alt,
    is_notification: isNotification,
    problems,
    payload,
  };
}

async function decodeQrCodes(imagePath) {
  // All QR payloads in an image via zbarimg. Returns list of strings.
  let stdout;
  try {
    ({ stdout } = await execFileP(
      "zbarimg",
      ["--raw", "-q", "-Sdisable", "-Sqrcode.enable", "-Sbinary", imagePath],
      { encoding: "buffer", maxBuffer: 16 << 20 }));
  } catch (e) {
    if (e.code === 4 && e.stdout) stdout = e.stdout;   // 4 = no symbols found
    else return [];
  }
  // -Sbinary stops zbar from charset-guessing (it mangles UTF-8 umlauts
  // otherwise); we decode UTF-8 ourselves. SPC payloads embed newlines,
  // so symbols are split on the 'SPC' header anchor, not on lines.
  const out = stdout.toString("utf-8");
  if (!out.trim()) return [];
  const idxs = [];
  for (const m of out.matchAll(/^SPC\r?$/gm)) idxs.push(m.index);
  if (idxs.length) {
    idxs.push(out.length);
    const chunks = [];
    for (let i = 0; i < idxs.length - 1; i++)
      chunks.push(out.slice(idxs[i], idxs[i + 1]).replace(/^\n+|\n+$/g, ""));
    return chunks;
  }
  return out.replace(/^\n+|\n+$/g, "").split("\n").filter(Boolean);
}

async function findQrbill(imagePaths) {
  // Scan page images for a Swiss QR-bill; first valid SPC wins.
  for (const p of imagePaths) {
    for (const payload of await decodeQrCodes(p)) {
      const parsed = parseSpc(payload);
      if (parsed) return parsed;
    }
  }
  return null;
}

async function renderQrPng(payload) {
  // Swiss QR PNG with the 7/46 Swiss-cross overlay per the IG (ECC level M
  // absorbs it) -- qrencode CLI + sharp SVG composite.
  const sharp = require("sharp");
  const { stdout } = await execFileP(
    "qrencode", ["-t", "PNG", "-l", "M", "-s", "10", "-m", "4", "-o", "-"],
    { encoding: "buffer", input: payload, maxBuffer: 16 << 20 });
  const img = sharp(stdout);
  const { width } = await img.metadata();
  const s = Math.round(width * 7 / 46);
  const stroke = Math.max(1, Math.round(s / 24));
  const arm = Math.round(s * 0.58), thick = Math.round(s * 0.18);
  const c = s / 2;
  const cross = Buffer.from(
    `<svg width="${s}" height="${s}" xmlns="http://www.w3.org/2000/svg">
       <rect x="0" y="0" width="${s}" height="${s}" fill="black"
             stroke="white" stroke-width="${stroke}"/>
       <rect x="${c - thick / 2}" y="${c - arm / 2}" width="${thick}" height="${arm}" fill="white"/>
       <rect x="${c - arm / 2}" y="${c - thick / 2}" width="${arm}" height="${thick}" fill="white"/>
     </svg>`);
  return img.composite([{ input: cross, gravity: "centre" }]).png().toBuffer();
}

const buildSpc = (qr) => qr.payload ?? null;

module.exports = { QR_IID_MIN, QR_IID_MAX, MOD10_TABLE, NOTIFICATION_TEXTS,
                   ibanValid, isQrIban, qrrValid, scorValid,
                   parseSwico, parseSpc, decodeQrCodes, findQrbill,
                   renderQrPng, buildSpc };
